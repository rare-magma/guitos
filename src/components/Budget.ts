import ItemForm from "./ItemForm";
import Stats from "./Stats";

type Items = typeof ItemForm[];

export class Budget {
  id!: number;
  name!: string;
  expenses: Items = [];
  incomes: Items = [];
  stats!: Stats;
  total = 0;
  get isNew(): boolean {
    return this.id === undefined;
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  constructor(initializer?: any) {
    if (!initializer) return;
    if (initializer.id) this.id = initializer.id;
    if (initializer.name) this.name = initializer.name;
    if (initializer.expenses) this.expenses = initializer.expenses;
    if (initializer.incomes) this.incomes = initializer.incomes;
    if (initializer.stats) this.stats = initializer.stats;
    if (initializer.total) this.total = initializer.total;
  }
}
