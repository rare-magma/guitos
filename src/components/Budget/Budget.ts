import { Stat } from "../StatCard/Stat";
import { Expense } from "../TableCard/Expense";
import { Income } from "../TableCard/Income";

export class Budget {
  id!: string;
  name!: string;
  expenses!: Expense;
  incomes!: Income;
  stats!: Stat;

  constructor(initializer?: Budget) {
    if (!initializer) return;
    if (initializer.id) this.id = initializer.id;
    if (initializer.name) this.name = initializer.name;
    if (initializer.expenses) this.expenses = initializer.expenses;
    if (initializer.incomes) this.incomes = initializer.incomes;
    if (initializer.stats) this.stats = initializer.stats;
  }
}
