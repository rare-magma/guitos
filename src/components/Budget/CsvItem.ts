export class CsvItem {
  type!: "expense" | "income" | "goal" | "reserves";
  name!: string;
  value!: number;

  constructor(initializer?: CsvItem) {
    if (!initializer) return;
    if (initializer.type) this.type = initializer.type;
    if (initializer.name) this.name = initializer.name;
    if (initializer.value) this.value = initializer.value;
  }
}
