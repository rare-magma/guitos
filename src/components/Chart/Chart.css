.recharts-default-tooltip,
.recharts-tooltip-wrapper {
  background-color: var(--lightbgcolor);
  border: 1px solid var(--comment);
}
