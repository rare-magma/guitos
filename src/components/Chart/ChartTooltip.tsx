import Big from "big.js";
import { Col, Container, Row } from "react-bootstrap";
import { useConfig } from "../../context/ConfigContext";
import { intlFormat, roundBig } from "../../utils";

interface ChartTooltipProps {
  active?: boolean;
  payload?: {
    name: string;
    payload?: {
      id: string;
      item: string;
      name: string;
      type: string;
      value: number;
    };
    value: number;
    unit: string;
  }[];
  label?: string;
  key1?: string | undefined;
  key2?: string | undefined;
}

function ChartTooltip({
  active,
  payload,
  label,
  key1,
  key2,
}: ChartTooltipProps) {
  const { intlConfig } = useConfig();
  const currency = intlConfig?.currency;
  const showTooltip = active && payload?.length && currency;
  const filteredItemName = payload?.length && payload[0].payload?.item;
  const showMultipleLegends = showTooltip && payload.length > 1;

  const item1Value =
    showTooltip && intlFormat(roundBig(Big(payload[0].value), 2), currency);
  const item2Value =
    showMultipleLegends &&
    intlFormat(roundBig(Big(payload[1].value), 2), currency);

  return showTooltip ? (
    <Container className="m-2">
      <Row>{label}</Row>
      {showMultipleLegends ? (
        <>
          <Row className="me-1">
            <Col>{`${key1 ?? ""}:`}</Col>
            <Col className="text-end fixed-width-font">{item1Value}</Col>
          </Row>
          <Row className="me-1 flex-md-nowrap">
            <Col>{`${key2 ?? ""}:`}</Col>
            <Col className="text-end fixed-width-font col-md-auto">
              {item2Value}
            </Col>
          </Row>
        </>
      ) : (
        <Row className="me-1">
          <Col>{`${key1 ?? filteredItemName}:`}</Col>
          {key1 === "goal" ? (
            <Col className="text-end fixed-width-font">{`${payload[0].value}%`}</Col>
          ) : (
            <Col className="text-end fixed-width-font">{item1Value}</Col>
          )}
        </Row>
      )}
    </Container>
  ) : null;
}

export default ChartTooltip;
