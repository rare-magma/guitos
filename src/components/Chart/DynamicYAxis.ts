/* eslint-disable eqeqeq */
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-unsafe-argument */
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-return */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
import { useCallback, useMemo, useState } from "react";

const RECHART_CERTESIAN_AXIS_TICK_VALUE_SELECTOR = `.recharts-cartesian-axis-tick-value[orientation="left"],
.recharts-cartesian-axis-tick-value[orientation="right"]`;

interface Props {
  yAxisWidthModifier?: (width: number) => number;
}

interface ReturnValues {
  yAxisWidth: undefined | number;
  setChartRef: (chartRef: any) => void;
}

function useDynamicYAxisWidth(props: void | Props): ReturnValues {
  const { yAxisWidthModifier } = props || {};
  const [yAxisWidthState, setYAxisWidthState] = useState(undefined);

  const setChartRef = useCallback(
    (chartRef: any) => {
      if (chartRef?.container != null) {
        const tickValueElements = chartRef.container.querySelectorAll(
          RECHART_CERTESIAN_AXIS_TICK_VALUE_SELECTOR,
        );
        const highestWidth = [...tickValueElements]
          .map((el) => {
            const boundingRect = el.getBoundingClientRect();
            if (boundingRect?.width != null) {
              return boundingRect.width;
            }
            return 0;
          })
          .reduce((accumulator, value) => {
            if (accumulator < value) {
              return value;
            }
            return accumulator;
          }, 0);
        setYAxisWidthState(highestWidth);
      }
    },
    [setYAxisWidthState],
  );

  const yAxisWidth = useMemo(() => {
    if (yAxisWidthModifier != null && yAxisWidthState != null) {
      return yAxisWidthModifier(yAxisWidthState);
    }
    return yAxisWidthState;
  }, [yAxisWidthModifier, yAxisWidthState]);

  return {
    yAxisWidth,
    setChartRef,
  };
}

export default useDynamicYAxisWidth;
