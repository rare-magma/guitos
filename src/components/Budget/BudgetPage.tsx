import Papa from "papaparse";
import { Suspense, lazy, useEffect, useRef, useState } from "react";
import { Col, Container, Row, ToastContainer } from "react-bootstrap";
import { useHotkeys } from "react-hotkeys-hook";
import { useParams } from "react-router-dom";
import { useBudget } from "../../context/BudgetContext";
import { useConfig } from "../../context/ConfigContext";
import { budgetsDB, optionsDB } from "../../context/db";
import {
  budgetToCsv,
  calcAutoGoal,
  calcAvailable,
  calcSaved,
  calcWithGoal,
  convertCsvToBudget,
  createBudgetNameList,
  createNewBudget,
  roundBig,
  userLang,
} from "../../utils";
import ErrorModal, { CsvError, JsonError } from "../ErrorModal/ErrorModal";
import LandingPage from "../LandingPage/LandingPage";
import Loading from "../Loading/Loading";
import NavBar, { SearchOption } from "../NavBar/NavBar";
import Notification from "../Notification/Notification";
import { Stat } from "../StatCard/Stat";
import StatCard from "../StatCard/StatCard";
import { Expense } from "../TableCard/Expense";
import { Income } from "../TableCard/Income";
import TableCard from "../TableCard/TableCard";
import { Budget } from "./Budget";
// import { useWhatChanged } from "@simbathesailor/use-what-changed";

const ChartsPage = lazy(() => import("../ChartsPage/ChartsPage"));

function BudgetPage() {
  const inputRef = useRef<HTMLInputElement>(null);

  const [loading, setLoading] = useState(true);
  const [showGraphs, setShowGraphs] = useState(false);
  const [error, setError] = useState<string | null>(null);
  const [csvError, setCsvError] = useState<CsvError[]>([]);
  const [jsonError, setJsonError] = useState<JsonError[]>([]);
  const [show, setShow] = useState(false);
  const [focus, setFocus] = useState("");

  const [notifications, setNotifications] = useState<
    {
      show: boolean;
      id?: string;
      body?: string;
    }[]
  >([]);

  const { budget, setBudget, budgetList, setBudgetList, setBudgetNameList } =
    useBudget();
  const params = useParams();
  const name = String(params.name);

  const showCards = !loading && !showGraphs && budget?.id;

  const { setIntlConfig, handleCurrency } = useConfig();

  useHotkeys("escape", (e) => !e.repeat && setNotifications([]), {
    preventDefault: true,
  });
  useHotkeys("s", (e) => !e.repeat && handleExportJSON(), {
    preventDefault: true,
  });
  useHotkeys("d", (e) => !e.repeat && handleExportCSV(), {
    preventDefault: true,
  });
  useHotkeys("a", (e) => !e.repeat && !showGraphs && handleNew(), {
    preventDefault: true,
  });
  useHotkeys("c", (e) => !e.repeat && !showGraphs && handleClone(), {
    preventDefault: true,
  });
  useHotkeys(
    "i",
    (e) => !e.repeat && !showGraphs && budget && setShowGraphs(true),
    {
      preventDefault: true,
    },
  );

  function handleError(e: unknown) {
    if (e instanceof Error) setError(e.message);
    setShow(true);
  }

  function handleIncomeChange(item: Income) {
    let newBudget: Budget | undefined;
    if (budget) {
      newBudget = budget;
      newBudget.incomes = item;
      newBudget.stats.available = roundBig(calcAvailable(newBudget), 2);
      newBudget.stats.withGoal = calcWithGoal(newBudget);
      newBudget.stats.saved = calcSaved(newBudget);

      setBudget({
        ...budget,
        incomes: {
          ...budget.incomes,
          items: item.items,
          total: item.total,
        },
        stats: {
          ...budget.stats,
          available: newBudget.stats.available,
          withGoal: newBudget.stats.withGoal,
          saved: newBudget.stats.saved,
        },
      });
    }
  }

  function handleExpenseChange(item: Expense) {
    let newBudget: Budget;
    if (budget) {
      newBudget = budget;
      newBudget.expenses = item;
      newBudget.stats.available = roundBig(calcAvailable(newBudget), 2);
      newBudget.stats.withGoal = calcWithGoal(newBudget);
      newBudget.stats.saved = calcSaved(newBudget);

      setBudget({
        ...budget,
        expenses: {
          ...budget.expenses,
          items: item.items,
          total: item.total,
        },
        stats: {
          ...budget.stats,
          available: newBudget.stats.available,
          withGoal: newBudget.stats.withGoal,
          saved: newBudget.stats.saved,
        },
      });
    }
  }

  function handleStatChange(item: Stat | undefined) {
    let newBudget: Budget;
    if (budget && item) {
      newBudget = budget;
      newBudget.stats = item;
      newBudget.stats.available = roundBig(calcAvailable(newBudget), 2);
      newBudget.stats.withGoal = calcWithGoal(newBudget);
      newBudget.stats.saved = calcSaved(newBudget);

      setBudget({
        ...budget,
        stats: {
          ...budget.stats,
          available: newBudget.stats.available,
          withGoal: newBudget.stats.withGoal,
          saved: newBudget.stats.saved,
          goal: item.goal,
          reserves: item.reserves,
        },
      });
    }
  }

  function handleAutoGoal(item: Stat | undefined) {
    let newBudget: Budget;
    if (budget && item) {
      newBudget = budget;
      newBudget.stats = item;
      newBudget.stats.goal = calcAutoGoal(budget);
      newBudget.stats.available = roundBig(calcAvailable(newBudget), 2);
      newBudget.stats.withGoal = calcWithGoal(newBudget);
      newBudget.stats.saved = calcSaved(newBudget);

      setBudget({
        ...budget,
        stats: {
          ...budget.stats,
          available: newBudget.stats.available,
          withGoal: newBudget.stats.withGoal,
          saved: newBudget.stats.saved,
          goal: newBudget.stats.goal,
          reserves: newBudget.stats.reserves,
        },
      });
    }
  }

  function handleRename(newName?: string | null) {
    let newBudget: Budget;
    if (budget && newName) {
      newBudget = {
        ...budget,
        name: newName,
      };
      setBudget(newBudget);
    }
  }

  function handleNew() {
    const newBudget = createNewBudget();

    let newBudgetList: Budget[] = [];
    if (budgetList) {
      newBudgetList = budgetList.concat(newBudget);
    } else {
      newBudgetList = newBudgetList.concat(newBudget);
    }

    setBudget(newBudget);
    setBudgetList(newBudgetList);
    setBudgetNameList(createBudgetNameList(newBudgetList));

    setNotifications([
      ...notifications,
      {
        show: true,
        id: crypto.randomUUID(),
        body: `created "${newBudget.name}" budget`,
      },
    ]);
  }

  function handleClone() {
    if (budget) {
      const newBudget = {
        ...budget,
        id: crypto.randomUUID(),
        name: budget.name + "-clone",
      };

      let newBudgetList: Budget[] = [];
      if (budgetList) {
        newBudgetList = budgetList.concat(newBudget);
      } else {
        newBudgetList = newBudgetList.concat(newBudget);
      }

      setNotifications([
        ...notifications,
        {
          body: `cloned "${budget.name}" budget`,
          id: crypto.randomUUID(),
          show: true,
        },
      ]);
      setBudget(newBudget);
      setBudgetList(newBudgetList);
      setBudgetNameList(createBudgetNameList(newBudgetList));
    }
  }

  function handleRemove(toBeDeleted: string) {
    budgetList &&
      budgetsDB
        .removeItem(toBeDeleted)
        .then(() => {
          const newBudgetList = budgetList
            .filter((item: Budget) => item.id !== toBeDeleted)
            .sort((a, b) => a.name.localeCompare(b.name))
            .reverse();

          setBudgetList(newBudgetList);
          setBudgetNameList(
            createBudgetNameList(newBudgetList as unknown as Budget[]),
          );

          setNotifications([
            ...notifications,
            {
              body: `deleted "${budget?.name}" budget`,
              id: crypto.randomUUID(),
              show: true,
            },
          ]);

          if (newBudgetList.length >= 1) {
            setBudget(newBudgetList[0]);
          } else {
            setBudget(undefined);
          }
        })
        .catch((e: unknown) => {
          handleError(e);
        });
  }

  function handleSelect(selectedBudget: SearchOption[] | undefined) {
    if (selectedBudget && budgetList) {
      const filteredList = budgetList.filter(
        (item: Budget) => item.id === selectedBudget[0].id,
      );

      filteredList && setBudget(filteredList[0]);

      if (selectedBudget[0].item && selectedBudget[0].item.length > 0) {
        setFocus(selectedBudget[0].item);
      }
    }
  }

  useEffect(() => {
    const element = document.querySelector(`input[value="${focus}"]`);
    if (element !== null) {
      (element as HTMLElement).focus();
    }
  }, [focus]);

  function handleGo(step: number, limit: number) {
    const sortedList = budgetList?.sort((a, b) => a.name.localeCompare(b.name));
    if (budget) {
      const index = sortedList?.findIndex((b) => b.name.includes(budget.name));
      if (index !== limit && sortedList) {
        handleSelect([
          sortedList[(index ?? 0) + step] as unknown as SearchOption,
        ]);
      }
    }
  }

  function handleGoHome() {
    if (budget) {
      const name = new Date().toISOString();
      const index = budgetList?.findIndex((b) =>
        b.name.includes(name.slice(0, 7)),
      );
      const isSelectable = index !== -1 && budgetList && index;

      if (isSelectable) {
        handleSelect([budgetList[index] as unknown as SearchOption]);
      }
    }
  }

  function handleGoBack() {
    budgetList && handleGo(-1, 0);
  }

  function handleGoForward() {
    budgetList && handleGo(1, budgetList.length - 1);
  }

  function handleImportCsv(fileReader: FileReader, file: File) {
    const newBudgetList: Budget[] = [];
    const csvObject = Papa.parse(fileReader.result as string, {
      header: true,
      skipEmptyLines: "greedy",
    });

    const hasErrors = csvObject.errors.length > 0;

    if (hasErrors) {
      csvError.push({
        errors: csvObject.errors,
        file: file.name,
      });
      setCsvError(csvError);
      setShow(true);
      setLoading(false);

      return;
    }

    const newBudget = convertCsvToBudget(
      csvObject.data as string[],
      file.name.slice(0, -4),
    );
    newBudgetList.push(newBudget);
    save(newBudget);
    setBudgetList(newBudgetList);
    setBudgetNameList(createBudgetNameList(newBudgetList));
  }

  function handleImportJSON(fileReader: FileReader, file: File) {
    const newBudgetList: Budget[] = [];
    try {
      const list = JSON.parse(fileReader.result as string) as Budget[];
      list.forEach((b: Budget) => {
        newBudgetList.push(b);
        save(b);
      });
      setBudgetList(newBudgetList);
      setBudgetNameList(createBudgetNameList(newBudgetList));
    } catch (e) {
      setJsonError([{ errors: (e as string).toString(), file: file.name }]);
      setShow(true);
      setLoading(false);
    }
  }

  function handleImport(e: React.ChangeEvent<HTMLInputElement>) {
    setLoading(true);
    const importedFiles = e.target.files;
    if (importedFiles === null) {
      return;
    }
    for (const file of importedFiles) {
      const reader = new FileReader();
      reader.readAsText(file, "UTF-8");
      reader.onloadend = () => {
        if (reader.result !== null) {
          if (file.type === "text/csv") {
            handleImportCsv(reader, file);
          } else {
            handleImportJSON(reader, file);
          }
        }
      };
    }
  }

  function handleExport(t: string) {
    if (t === "csv") handleExportCSV();
    if (t === "json") handleExportJSON();
  }

  function handleExportJSON() {
    if (budget) {
      let filename = new Date().toISOString();
      filename = `guitos-${filename.slice(0, -5)}.json`;
      const url = window.URL.createObjectURL(
        new Blob([JSON.stringify(budgetList)]),
      );
      const link = document.createElement("a");
      link.href = url;
      link.setAttribute("download", filename);
      document.body.appendChild(link);
      link.click();
    }
  }

  function handleExportCSV() {
    if (budget) {
      const filename = `${budget.name}.csv`;

      const url = window.URL.createObjectURL(new Blob([budgetToCsv(budget)]));
      const link = document.createElement("a");
      link.href = url;
      link.setAttribute("download", filename);
      document.body.appendChild(link);
      link.click();
    }
  }

  function save(budget: Budget) {
    let list: Budget[] = [];
    budgetsDB
      .setItem(budget.id, budget)
      .then(() => {
        budgetsDB
          .iterate((value) => {
            list = list.concat(value as Budget);
          })
          .then(() => {
            setBudgetList(list);
            setBudgetNameList(createBudgetNameList(list));
          })
          .catch((e: unknown) => {
            handleError(e);
          });
      })
      .catch((e: unknown) => {
        handleError(e);
      });
  }

  function loadFromDb() {
    let list: Budget[] = [];

    budgetsDB
      .iterate((value) => {
        list = list.concat(value as Budget);
      })
      .then(() => {
        setBudgetList(list);
        setBudgetNameList(createBudgetNameList(list));

        if (name.trim() !== "undefined") {
          setBudget(list.filter((b: Budget) => b && b.name === name)[0]);
        } else {
          setBudget(
            list
              .sort((a, b) => a.name.localeCompare(b.name))
              .reverse()
              .filter((b: Budget) => b && b.id === list[0].id)[0],
          );
        }

        loadCurrencyOption();
        setLoading(false);
      })
      .catch((e) => {
        handleError(e);
      });
  }

  function loadBudget(list: Budget[]) {
    list.forEach((data: Budget) => {
      budgetsDB
        .getItem(data.id)
        .then((b) => setBudget(b as Budget))
        .catch((e) => {
          handleError(e);
        });
    });
  }

  function loadCurrencyOption() {
    optionsDB
      .getItem("currencyCode")
      .then((c) => {
        if (c) {
          handleCurrency(c as string);
          setIntlConfig({ locale: userLang, currency: c as string });
        }
      })
      .catch((e) => {
        handleError(e);
      });
  }

  // useWhatChanged([budget, name]);

  useEffect(() => {
    budget && save(budget);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [budget]);

  useEffect(() => {
    try {
      const shouldLoadBudgetsFromList =
        budgetList && budgetList.length >= 1 && Array.isArray(budgetList);

      if (shouldLoadBudgetsFromList) {
        if (name.trim() !== "undefined") {
          loadBudget(budgetList.filter((b: Budget) => b && b.name === name));
        } else {
          loadBudget(budgetList.slice(0));
        }

        loadCurrencyOption();
        setBudgetNameList(createBudgetNameList(budgetList));
        setLoading(false);
      } else {
        loadFromDb();
      }
    } catch (e: unknown) {
      handleError(e);
      setLoading(false);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [name, loading]);

  return (
    <Container fluid style={{ zIndex: 1 }}>
      <ToastContainer
        className="p-2"
        position={"bottom-center"}
        style={{ zIndex: 100 }}
      >
        {notifications.map((notification) => {
          return (
            notification && (
              <Notification
                notification={notification}
                onShow={() => {
                  setNotifications(
                    notifications.filter((n) => n.id !== notification.id),
                  );
                }}
              />
            )
          );
        })}
      </ToastContainer>

      {!showGraphs && (
        <NavBar
          onRename={(e) => {
            handleRename(e);
          }}
          onClone={() => {
            handleClone();
          }}
          onExport={(t) => {
            handleExport(t);
          }}
          onGoBack={() => {
            handleGoBack();
          }}
          onGoHome={() => {
            handleGoHome();
          }}
          onGoForward={() => {
            handleGoForward();
          }}
          onNew={() => {
            handleNew();
          }}
          onImport={(e) => {
            handleImport(e);
          }}
          onRemove={(e) => {
            handleRemove(e);
          }}
          onSelect={(e) => {
            handleSelect(e);
          }}
        />
      )}

      <LandingPage
        loading={loading}
        inputRef={inputRef}
        onNew={handleNew}
        onImport={(e) => {
          handleImport(e);
        }}
      />

      <ErrorModal
        error={error}
        show={show}
        jsonError={jsonError}
        csvError={csvError}
        onShow={setShow}
        onError={() => {
          setJsonError([]);
          setCsvError([]);
        }}
      />

      {showGraphs && (
        <Suspense fallback={<Loading />}>
          <ChartsPage onShowGraphs={() => setShowGraphs(false)} />
        </Suspense>
      )}

      {showCards && (
        <Container key={budget.id}>
          <Row className="mt-1">
            <Col md="6">
              <div className="card-columns">
                <StatCard
                  key={budget?.expenses.total + budget?.incomes.total}
                  onChange={handleStatChange}
                  onAutoGoal={handleAutoGoal}
                  onShowGraphs={() => setShowGraphs(true)}
                />

                <div className="mt-3" />

                <TableCard
                  items={budget.incomes}
                  header="Revenue"
                  onChange={handleIncomeChange}
                />
                <div className="mt-3" />
              </div>
            </Col>

            <Col md="6" className="mb-3">
              <TableCard
                items={budget.expenses}
                header="Expenses"
                onChange={handleExpenseChange}
              />
            </Col>
          </Row>
        </Container>
      )}
    </Container>
  );
}

export default BudgetPage;
