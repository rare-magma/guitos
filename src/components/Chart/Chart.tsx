import { Card, Col, Form, InputGroup, Row } from "react-bootstrap";
import CurrencyInput from "react-currency-input-field";
import { BsPercent } from "react-icons/bs";
import {
  Area,
  AreaChart,
  ResponsiveContainer,
  Tooltip,
  XAxis,
  YAxis,
} from "recharts";
import { useBudget } from "../../context/BudgetContext";
import { useConfig } from "../../context/ConfigContext";
import { intlFormat, median } from "../../utils";
import { FilteredItem } from "../ChartsPage/ChartsPage";
import "./Chart.css";
import { ChartTooltip } from "./ChartTooltip";
import { useDynamicYAxisWidth } from "./DynamicYAxis";

interface ChartProps {
  header: string;
  tooltipKey1?: string;
  tooltipKey2?: string;
  areaDataKey1: string;
  areaDataKey2?: string;
  areaStroke1: string;
  areaFill1: string;
  areaStroke2?: string;
  areaFill2?: string;
  legend1: string;
  legendValues1: number[];
  legend2?: string;
  legendValues2?: number[];
  unit?: string;
  filteredData?: FilteredItem[];
}

const horizonalRatio = 3.4;
const verticalRatio = 1.6;

export function Chart({
  header,
  tooltipKey1,
  tooltipKey2,
  areaDataKey1,
  areaDataKey2,
  areaStroke1,
  areaStroke2,
  areaFill1,
  areaFill2,
  legend1,
  legend2,
  legendValues1,
  legendValues2,
  unit,
  filteredData,
}: ChartProps) {
  const { budgetList } = useBudget();
  const { intlConfig } = useConfig();

  const showSecondArea = areaDataKey2 && areaStroke2 && areaFill2;
  const isGoalChart = tooltipKey1 === "goal";
  const isVerticalScreen = window.innerWidth < window.innerHeight;
  const chartData =
    filteredData ?? budgetList?.sort((a, b) => a.name.localeCompare(b.name));

  const { yAxisWidth, setChartRef } = useDynamicYAxisWidth({
    yAxisWidthModifier: (x) => x + 10,
  });

  function medianLabelGroup(legend: string, values: number[]) {
    return (
      <InputGroup size="sm" className="mb-1">
        <InputGroup.Text>{legend}</InputGroup.Text>
        <CurrencyInput
          disabled
          className="text-end form-control fixed-width-font"
          aria-label={legend}
          intlConfig={intlConfig}
          defaultValue={median(values)}
        />
      </InputGroup>
    );
  }

  function tickFormatter(value: number) {
    return (
      (intlConfig?.currency && intlFormat(value, intlConfig.currency)) ?? ""
    );
  }

  return (
    <Card className="stat-card mt-3 d-flex flex-grow-1">
      <Card.Header className="stat-card-header">{header}</Card.Header>
      <Card.Body>
        <ResponsiveContainer
          width="100%"
          aspect={isVerticalScreen ? verticalRatio : horizonalRatio}
        >
          <AreaChart
            data={chartData}
            ref={setChartRef}
            margin={{
              top: 10,
              right: 0,
              bottom: 0,
            }}
          >
            <XAxis
              stroke="var(--textcolor)"
              dataKey="name"
              minTickGap={10}
              tick={{ transform: "translate(0, 10)" }}
            />
            {unit ? (
              <YAxis
                stroke="var(--textcolor)"
                style={{ fontFamily: "ui-monospace, monospace" }}
                width={yAxisWidth}
                unit={unit}
              />
            ) : (
              <YAxis
                stroke="var(--textcolor)"
                style={{ fontFamily: "ui-monospace, monospace" }}
                width={yAxisWidth}
                tickFormatter={tickFormatter}
              />
            )}
            <Tooltip
              allowEscapeViewBox={{ x: false, y: false }}
              position={{ y: 0 }}
              content={<ChartTooltip key1={tooltipKey1} key2={tooltipKey2} />}
              contentStyle={{ backgroundColor: "var(--bgcolor)" }}
              itemStyle={{ color: "var(--textcolor)" }}
            />
            <Area
              type="monotone"
              dataKey={areaDataKey1}
              stroke={`var(--${areaStroke1})`}
              fill={`var(--${areaFill1})`}
              isAnimationActive={false}
            />
            {showSecondArea && (
              <Area
                type="monotone"
                dataKey={areaDataKey2}
                stroke={`var(--${areaStroke2})`}
                fill={`var(--${areaFill2})`}
                isAnimationActive={false}
              />
            )}
          </AreaChart>
        </ResponsiveContainer>
      </Card.Body>
      <Card.Footer className="stat-card-footer">
        <Row>
          {isGoalChart && (
            <InputGroup size="sm" className="mb-1">
              <InputGroup.Text>{legend1}</InputGroup.Text>
              <Form.Control
                className="text-end fixed-width-font"
                aria-label={"median"}
                defaultValue={legendValues1 && median(legendValues1)}
                type="number"
                disabled
              />
              <InputGroup.Text>
                <BsPercent aria-hidden />
              </InputGroup.Text>
            </InputGroup>
          )}
          {!isGoalChart &&
            !legendValues2 &&
            medianLabelGroup(legend1, legendValues1)}
          {!isGoalChart && legendValues2 && (
            <Col lg={6}>{medianLabelGroup(legend1, legendValues1)}</Col>
          )}
          {legend2 && legendValues2 && (
            <Col>{medianLabelGroup(legend2, legendValues2)}</Col>
          )}
        </Row>
      </Card.Footer>
    </Card>
  );
}
