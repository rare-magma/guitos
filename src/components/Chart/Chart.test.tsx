import { render, screen } from "@testing-library/react";
import { testIntlConfig, testBudgetList } from "../../setupTests";
import Chart from "./Chart";
import { Budget } from "../Budget/Budget";
import { vi } from "vitest";

describe("Chart", () => {
  const comp = (
    <Chart
      header={"chart header"}
      budgetList={testBudgetList}
      intlConfig={testIntlConfig}
      tooltipKey1={"tooltipKey1"}
      areaDataKey1={"revenue"}
      areaStroke1={"highlight"}
      areaFill1={"highlight"}
      legend1={"median revenue"}
      legendValues1={testBudgetList.map((b: Budget) => {
        return b.incomes.total;
      })}
    />
  );
  beforeAll(() => {
    vi.spyOn(HTMLElement.prototype, "clientHeight", "get").mockReturnValue(800);
    vi.spyOn(HTMLElement.prototype, "clientWidth", "get").mockReturnValue(800);
  });

  beforeEach(() => {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    //@ts-ignore
    delete window.ResizeObserver;
    window.ResizeObserver = vi.fn().mockImplementation(() => ({
      observe: vi.fn(),
      unobserve: vi.fn(),
      disconnect: vi.fn(),
    }));

    render(comp);
  });

  afterEach(() => {
    window.ResizeObserver = ResizeObserver;
    vi.restoreAllMocks();
  });

  it("matches snapshot", () => {
    expect(comp).toMatchSnapshot();
  });
  it("renders initial state", () => {
    expect(screen.getByText("chart header")).toBeInTheDocument();
    expect(screen.getByText("median revenue")).toBeInTheDocument();
    expect(screen.getByDisplayValue("$200")).toBeInTheDocument();
  });
});
