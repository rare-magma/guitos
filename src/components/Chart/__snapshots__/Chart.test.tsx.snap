// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Chart > matches snapshot 1`] = `
<Chart
  areaDataKey1="revenue"
  areaFill1="highlight"
  areaStroke1="highlight"
  budgetList={
    [
      Budget {
        "expenses": {
          "items": [
            {
              "id": 1,
              "name": "expense1",
              "value": 10,
            },
          ],
          "total": 10,
        },
        "id": "035c2de4-00a4-403c-8f0e-f81339be9a4e",
        "incomes": {
          "items": [
            {
              "id": 2,
              "name": "income1",
              "value": 100,
            },
          ],
          "total": 100,
        },
        "name": "2023-03",
        "stats": {
          "available": 90,
          "goal": 10,
          "reserves": 200,
          "saved": 10,
          "withGoal": 80,
        },
      },
      Budget {
        "expenses": {
          "items": [
            {
              "id": 1,
              "name": "name",
              "value": 50,
            },
          ],
          "total": 50,
        },
        "id": "135b2ce4-00a4-403c-8f0e-f81339be9a4e",
        "incomes": {
          "items": [
            {
              "id": 2,
              "name": "name",
              "value": 200,
            },
          ],
          "total": 200,
        },
        "name": "2023-04",
        "stats": {
          "available": 150,
          "goal": 35,
          "reserves": 30,
          "saved": 20,
          "withGoal": 130,
        },
      },
      Budget {
        "expenses": {
          "items": [
            {
              "id": 1,
              "name": "name",
              "value": 11378.64,
            },
            {
              "id": 4,
              "name": "name2",
              "value": 11378.64,
            },
          ],
          "total": 22757.28,
        },
        "id": "035c2de4-00a4-403c-8f0e-f81339be9a4e",
        "incomes": {
          "items": [
            {
              "id": 2,
              "name": "name",
              "value": 100.03,
            },
            {
              "id": 3,
              "name": "name2",
              "value": 342783.83,
            },
          ],
          "total": 342883.86,
        },
        "name": "2023-03",
        "stats": {
          "available": 320126.58,
          "goal": 50,
          "reserves": 200,
          "saved": 171441.93,
          "withGoal": 148684.65,
        },
      },
    ]
  }
  header="chart header"
  intlConfig={
    {
      "currency": "USD",
      "locale": "en-US",
    }
  }
  legend1="median revenue"
  legendValues1={
    [
      100,
      200,
      342883.86,
    ]
  }
  tooltipKey1="tooltipKey1"
/>
`;
